Implements prereq command

sneakyteak opened the pull request:
Adds an extension of the boon command which responds to command syntax like !flutter flourish prereqs. Code satisfies new scenarios which test prereqs queries. Adds a new formatter to construct prereq strings. As prereq definitions ask almost every god to depend on almost every other god, it seemed important in testing for the formatter to not be called until runtime / in the handler. Attempts to add the strings to god definitions choked on incomplete data in the prereq arrays.

sneakyteak committed:
Implements prereq command

Adds an extension of the boon command which responds to command syntax
like !flutter flourish prereqs. Code satisfies new scenarios which test
prereqs queries. Adds a new formatter to construct prereq strings. As
prereq definitions ask almost every god to depend on almost every other
god, it seemed important in testing for the formatter to not be called
until runtime / in the handler. Attempts to add the strings to god
definitions choked on incomplete data in the prereq arrays.

Ellisande reviewed (approved):
I think the changes are fine. I was worried it would be a big command after your comments in disc, but I think its very reasonable. That being said I'm not sure I'm in love with `prereqs` as a user-facing term.  I don't have a great counter suggestion right now... 
!backBurner requires
!backBurner conditions
!backBurner needs

Or maybe we support all of them and let the user decide? Not sure, but I'm good with this if you are =)

Ellisande commented on features/support/bot_steps.ts:
```diff
@@ -77,7 +77,7 @@ Then(
   function (this: CustomWorld, expectedMessage: string) {
     const { bot } = this.mocks;
-    const calledWith = bot.say.lastArg;
+    const calledWith = bot.say.getCall(0).args;
```
I'm not necessarily surprised you had to do this, but very surprised it didn't break anything else

sneakyteak replied:
Yeah, the "does not respond" scenario wasn't working so I just copied what was working in the "does respond" definition.

sneakyteak commented:
> I think the changes are fine. I was worried it would be a big command after your comments in disc, but I think its very reasonable. That being said I'm not sure I'm in love with `prereqs` as a user-facing term. I don't have a great counter suggestion right now... !backBurner requires !backBurner conditions !backBurner needs
> 
> Or maybe we support all of them and let the user decide? Not sure, but I'm good with this if you are =)

Now that you mention it I agree that prereqs feels bad as a user facing term. As the regex is right now, it actually also supports 'reqs' which I think is reasonable. I like the 'requires' or 'needs' ideas. What language does the actual in-game codex use?

Ellisande commented:
Looks like the in-game codex revers to them as `Offering Requirements`

sneakyteak committed:
Updated prereqs user-facing language

Now uses 'reqs' or 'requirements' wherever user input is expected or a
response is provided to a user. Adds scenarios to test new user syntax
and to cover a previously untested case (asking for prereqs for a boon
that has none).
